ctpolicy/ctpolicy.go

```diff
@@ -49,13 +51,102 @@ func (group *LogGroupInfo) setMinInclusions(i int) error {
 
 func (group *LogGroupInfo) populate(ll *loglist.LogList, included func(log *loglist.Log) bool) {
 	group.LogURLs = make(map[string]bool)
+	group.LogWeights = make(map[string]float32)
 	for _, l := range ll.Logs {
 		if included(&l) {
 			group.LogURLs[l.URL] = true
+			group.LogWeights[l.URL] = 1.0
 		}
 	}
 }
 
+// resetLogWeights applies suggested weights to the Log-group. Does not reset
+// weights and returns error when not enough to reach minimal inclusion number
+// positive weights provided.
+func (group *LogGroupInfo) ResetLogWeights(weights map[string]float32) error {
+	groupWeights := make(map[string]float32, len(group.LogURLs))
+	for logURL := range group.LogURLs {
+		groupWeights[logURL] = 1.0
+	}
+	nonZeroNum := len(group.LogURLs)
+	for logURL, w := range weights {
+		if group.LogURLs[logURL] {
+			groupWeights[logURL] = w
+			if w <= 0 {
+				nonZeroNum--
+			}
+		}
+	}
+	if nonZeroNum < group.MinInclusions {
+		return fmt.Errorf("trying to assign weights %v resulting into unability to reach minimal inclusion number %d", weights, group.MinInclusions)
+	}
+	group.LogWeights = groupWeights
+	return nil
+}
+
+// setLogWeight tries setting weights for a single Log of the Log-group.
+// Does not reset the weight and returns error if weight is non-positive and
+// its setting will result innto unability to reach minimal inclusion number.
+func (group *LogGroupInfo) SetLogWeight(logURL string, w float32) error {
+	if !group.LogURLs[logURL] {
+		return nil
+	}
```

Comment:
should this fail if `w < 0` ?

Reply:
I treat negative weights as zero ones.

Reply:
It is probably better to be strict with the inputs and error. 
1) It will help catch any logic bugs earlier
2) It can avoid super unexpected behavior from underflows

Reply:
Restricted to positive weights.